app/components/graph-editor/drawable-interfaces.ts

```diff
@@ -0,0 +1,398 @@
+    clone.lineWidth = e.lineWidth;
+    clone.label = e.label;
+    return clone;
+}
+
+/**
+ * cloneNode  
+ *   Creates a cloned node.
+ */
+export function cloneNode(n: DrawableNode): DrawableNode {
+    let clone = new DefaultNode(n.x, n.y);
+    clone.label = n.label;
+    clone.shape = n.shape;
+    clone.color = n.color;
+    clone.borderColor = n.borderColor;
+    clone.borderStyle = n.borderStyle;
+    clone.borderWidth = n.borderWidth;
+    return clone;
+}
+
+
+// Interfaces //////////////////////////////////////////////////////////////////
+
+
+/**
+ * GraphContext  
+ *   Interface that exposes a graph and any external properties that the 
+ *   editor needs to have to go with it. Specifically the set of selected
+ *   `DrawableNode | DrawableEdge`
+ */
+export interface GraphContext {
+    readonly graph: DrawableGraph;
+    readonly selectedDrawables: Set<Drawable>;
+}
+
+/**
+ * DrawableGraph  
+ *   Interface that exposes drawable graph properties and methods.
+ */
+export interface DrawableGraph {
+
+    /**
+     * nodes  
+     *   The collection of drawable nodes that are part of the graph.
+     */
+    readonly nodes: Iterable<DrawableNode>;
+
+    /**
+     * edges  
+     *   The collection of drawable edges that are part of the graph.
+     */
+    readonly edges: Iterable<DrawableEdge>;
+
+    /**
+     * backgroundColor  
+     *   What is this doing here? It should be a public property of the graph
+     *   editor component.
+     */
+    backgroundColor: string;
+
+    /**
+     * createNode  
+     *   Creates a drawable node with an optional position.
+     */
+    createNode(x?: number, y?: number): DrawableNode;
+
+    /**
+     * removeNode  
+     *   Guarantees that the given node is not present in the graph.
+     */
+    removeNode(node: DrawableNode): void;
+
+    /**
+     * canCreateEdge  
+     *   Checks if a drawable edge can be created with the given source and
+     *   destination nodes.
+     * 
+     *   If `like` is specified, a drawable edge with a matching type of `like` is
+     *   is checked against the given source and destination nodes.
+     * 
+     *   This method must be called before calling `createEdge`.
+     */
+    canCreateEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): boolean;
+
+    /**
+     * createEdge  
+     *   Creates a drawable edge with a source and destination node.
+     * 
+     *   If `like` is specified, a drawable edge with a matching type of `like` is
+     *   created.
+     * 
+     *   The `canCreateEdge` method must be called to check if creating the edge
+     *   is valid.
+     */
+    createEdge(
+        src: DrawableNode,
+        dst: DrawableNode,
+        like?: DrawableEdge
+    ): DrawableEdge;
+
+    /**
+     * replaceEdge  
+     *   This method has been replaced by `createEdge` and is no longer needed.
+     */
+    replaceEdge(original: DrawableEdge, replacement: DrawableEdge): void;
```

Comment:
This should be removed since this is a refactor or the comment should be updated.